docs(ivy): document global debugging utilities and clean up API

crisbeto opened the pull request:
Cleans up the public API of the global debugging utilities, documents them and exposes them in the API docs.

mary-poppins commented:
You can preview 20210a8 at https://pr34453-20210a8.ngbuilds.io/.
You can preview 0a60c88 at https://pr34453-0a60c88.ngbuilds.io/.

kara commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -57,52 +46,29 @@ export function getComponent<T = {}>(element: Element): T|null {
- *       <child-comp></child-comp>
- *     </div>
- * </mp-app>
+ * Retrieves the instance of the component that an element is inside of.
```
What's the difference between this function and `getViewComponent`?  I think this needs an example that shows it.

crisbeto replied:
I wanted to raise this in the next meeting, because I'm not exactly sure what the difference is either. From my testing the return values were identical.

kara replied:
@mhevery ?

kara replied:
Talked to Misko offline about this and he reminded me of embedded views. This will return the context of the closest embedded view if the element is inside an embedded view. So maybe:

```suggestion
 * If inside an embedded view (e.g. *ngIf or *ngFor), retrieves the context of the embedded view that owns the element.
 * If not, retrieves the instance of the component that owns the element (in this case, the result should be the same as `getOwnerComponent`

```

kara commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -277,25 +251,13 @@ export function isBrowserEvents(listener: Listener): boolean {
  * @publicApi
+ * @globalApi ng
  */
 export function getListeners(element: Element): Listener[] {
```
It feels like we should also have docs for `Listener` since we are listing it as a return value. There doesn't seem to be a way to click through right now.

crisbeto replied:
My thinking when I decided not to document it was that the way people interact with these utilities is usually in the browser console where they won't have access to interfaces anyway.

kara replied:
It also means that the return value of this function is completely opaque. You can't click through, so people won't know what this function returns, except that it's an array of an object called `Listener[]`. I think we need to document it.

kara reviewed (changes requested):
Thanks for picking this up! Some thoughts below

crisbeto commented on packages/core/src/render3/util/discovery_utils.ts:
````diff
@@ -24,31 +24,20 @@ import {getTNode, unwrapRNode} from './view_utils';
- *
- * expect(getComponent(<child-comp>) instanceof ChildComponent).toBeTruthy();
- * expect(getComponent(<my-app>) instanceof MyApp).toBeTruthy();
- * ```
````
In general I dropped the examples, because we don't seem to have any other inline examples on AIO. Maybe @kapunahelewong, @gkalpak or @petebacondarwin can say what the convention is?

kara replied:
I think the problem is that the one line explanations are not sufficient without examples. Can we see what it would look like to re-add them with backticks?

kapunahelewong replied:
@crisbeto, you're right, we do usually try to have functioning examples, especially in the guides, but we do have some inline examples in cases where it's hypothetical or doesn't need the context of an app. I think the inline ones are more common in the api docs, but maybe @gkalpak or @petebacondarwin would know for sure. (I've worked more in the guides and less in the api docs).

crisbeto commented on packages/core/src/render3/instructions/change_detection.ts:
```diff
@@ -31,28 +31,25 @@ export function detectChanges<T>(component: T): void {
+ * @globalApi ng
  */
-export function markDirty<T>(component: T) {
+export function markDirty<T>(component: T): void {
```
I think I missed renaming it to `applyChangeDetection`. That being said, isn't applying change detection different from marking it as dirty? When I hear "apply change detection" I assume that change detection would be run immediately, but that's not the case here.

kara replied:
Yeah, agree that the name makes it sound like it should be synchronous and local (rather than scheduling an async application tick). It feels like if we want to expose anything, it really should be `detectChanges`.

kara replied:
Chatted a bit with @mhevery about this and we decided:
 - We shouldn't expose `markDirty` (remove it from packages/core/src/render3/global_utils_api.ts in a separate commit)
- We should expose `detectChanges` instead and document it here

kara commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -57,52 +46,29 @@ export function getComponent<T = {}>(element: Element): T|null {
  * @publicApi
+ * @globalApi ng
  */
 export function getViewComponent<T = {}>(element: Element | {}): T|null {
```
Re - Misko/I talked offline and we agreed on a compromise - `getOwnerComponent`

crisbeto committed:
docs(ivy): document global debugging utilities and clean up API

Cleans up the public API of the global debugging utilities, documents them and exposes them in the API docs.

crisbeto commented:
I've reworked it based on the feedback @kara.

mary-poppins commented:
You can preview 889cf5b at https://pr34453-889cf5b.ngbuilds.io/.

IgorMinar commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -119,26 +114,30 @@ export function getViewComponent<T = {}>(element: Element | {}): T|null {
+ * @globalApi ng
  */
-export function getInjector(target: {}): Injector {
+export function getInjector(target: Element | {}): Injector {
```
plz rename `target` to `element` so that we use consistent naming across all functions

crisbeto replied:
It's not just an element though, it can also be a directive/component instance.

kara replied:
Maybe `elementOrDir`?

IgorMinar commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -119,26 +114,30 @@ export function getViewComponent<T = {}>(element: Element | {}): T|null {
+ * @globalApi ng
  */
-export function getInjector(target: {}): Injector {
+export function getInjector(target: Element | {}): Injector {
```
do we need `| {}` here? why? maybe some of our internal tests need it today because of mocks, but that should not affect the public api which should be just `Element`. Please fix across all functions.

crisbeto replied:
The `| {}` here is to signal that component/directive instances are accepted, in addition to a DOM element.

IgorMinar commented on packages/core/src/render3/instructions/change_detection.ts:
```diff
@@ -15,44 +15,34 @@ import {detectChangesInternal, markViewDirty, scheduleTick, tickRootContext} fro
+ * @publicApi
+ * @globalApi ng
  */
 export function detectChanges<T>(component: T): void {
```
but more importantly is this the right function to expose? I discussed this briefly with @kara yesterday and we have two different types of users that need to trigger change detection:
- application developers - those almost always want to trigger global change detection including flushing microtask queue that might have been filled during change detection
- component developers - those might want to trigger local change detection only for a subset of a component tree (similar to [ComponentFixture#detectchanges](https://pr34453-889cf5b.ngbuilds.io/api/core/testing/ComponentFixture#detectchanges))

I argue that unlike in unit tests where ComponentFixture#detectchanges is handy, during debugging triggering local change detection is less important and in fact sometimes undesirable because it could put the app into an inconsistent state.

For these reasons I propose that we initially expose a function that addresses the application developer needs (triggers global change detection) and only if we get requests from component developers expose another one that is triggers local change detection. 

This application developer specific function, should be implemented similar to how `markDirty` works today by triggering a global change detection.

In fact we could go as far as implementing it as:
```
function applyChanges(component?: {}): void
```
where if `component` is provided, we mark it as dirty and trigger synchronous change detection on the root component the provided component belongs to.

if the `component` is not provided, we trigger synchronous change detection on all root components (retrieved via `getRootComponents`).

crisbeto replied:
I'm not sure what the action item is for this PR. Should I implement the new utility or leave it as is until we've had a chance to discuss it?

kara replied:
Let's implement `applyChanges`. I'm good with what Igor proposes above

kara replied:
Let's implement `applyChanges`. I'm good with what Igor proposes above

crisbeto replied:
I've added `applyChanges` based on what Igor mentioned above, but I'm not sure how the case where a `component` isn't provided will work since, from what I can tell, we don't have a utility that allows us to get all of the root components without a reference DOM node.

IgorMinar replied:
aha! that is interesting. let's move the discussion to `applyChanges` - I'll start a new thread there.

IgorMinar reviewed (changes requested):
added a few more comments.. sorry for all the feedback, but I think the API and the docs really could use a significant cleanup compared to the current state.

crisbeto commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -24,87 +24,82 @@ import {getTNode, unwrapRNode} from './view_utils';
  * @publicApi
+ * @globalApi ng
  */
 export function getComponent<T = {}>(element: Element): T|null {
```
I'll just change it now. There are 24 usages of it and none of them are breaking for now.

crisbeto commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -263,10 +277,18 @@ export function loadLContextFromNode(node: Node): LContext {
 
+/**
+ * Represents an event listener configuration returned from `getListeners`.
+ * @publicApi
```
I was thinking of exporting it through there, but it seemed wrong to export an interface through a package that is supposed to be only global functions available at runtime. I don't feel strongly against doing it, but it seems like it would stand out next to the rest of the functions.

kara replied:
Agree that we should probably just export it

crisbeto commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -263,10 +277,18 @@ export function loadLContextFromNode(node: Node): LContext {
+  /** Callback that is invoked when the event is triggered. */
   callback: (value: any) => any;
+  /** Whether the listener is using event capturing. */
   useCapture: boolean|null;
```
I agree that it looks weird, but from what I can tell the `null` is there so we can distinguish between browser events and outputs which is used later on inside `DebugNode.listeners`. I see a couple of alternatives:

1. Add a separate flag on `Listener` that indicates whether it's an output.
2. Exclude outputs from the `getListeners` result.

crisbeto reviewed (commented):
I've implemented or responded to all of the feedback @IgorMinar.

mary-poppins commented:
You can preview a055e5f at https://pr34453-a055e5f.ngbuilds.io/.

mary-poppins commented:
You can preview d2099ac at https://pr34453-d2099ac.ngbuilds.io/.

mary-poppins commented:
You can preview 0c4be0a at https://pr34453-0c4be0a.ngbuilds.io/.

kara commented on tools/public_api_guard/global_utils.d.ts:
```diff
@@ -1,19 +1,17 @@
-export declare function getRootComponents(target: {}): any[];
-
-export declare function getViewComponent<T = {}>(element: Element | {}): T | null;
+export declare function getOwnerComponent<T = {}>(element: Element | {}): T | null;
```
Ha, Misko and I couldn't decide which one of those two sounded better, so we ended up with `getOwnerComponent`. `getOwningComponent` is fine with me too.

mary-poppins commented:
You can preview 8a69c30 at https://pr34453-8a69c30.ngbuilds.io/.

crisbeto committed:
refactor(ivy): remove markDirty from global utilities

Removes `markDirty` from the global debugging utilities API since we've already exposed `detectChanges`.

mary-poppins commented:
You can preview 0a2840b at https://pr34453-0a2840b.ngbuilds.io/.

crisbeto commented:
@kara @IgorMinar, I've addressed the latest set of feedback, can you take another look?

IgorMinar commented on packages/core/src/render3/util/change_detection_utils.ts:
```diff
@@ -0,0 +1,23 @@
+ * @publicApi
+ * @globalApi ng
+ */
+export function applyChanges(component: {}): void {
```
continuing @crisbeto's discussion from markDirty:
>I'm not sure how the case where a component isn't provided will work since, from what I can tell, we don't have a utility that allows us to get all of the root components without a reference DOM node.

aha! that is interesting. Doesn't protractor have a way to get hold of all roots via https://next.angular.io/api/core/TestabilityRegistry ? could we use that and change detect all the root components?

crisbeto replied:
I looked into it and we should be able to use `TestabilityRegistry` to get all the root elements, but the problem is that we need an injector so that we can get a hold of a `TestabilityRegistry` in the first place, and we can't do that without a DOM node or a component instance.

IgorMinar replied:
I see. let's skip this for now. We might come back to this later, but it can be done after 9.0

IgorMinar reviewed (changes requested):
this is much better @crisbeto a few more changes requested but we are close. thanks!

crisbeto commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -230,17 +253,17 @@ export function getLocalRefs(target: {}): {[key: string]: any} {
+ * The host element is the DOM element that matched the selector of the directive.
  *
- * @param directive Component or Directive for which the host element should be retrieved.
+ * @param target Component or directive instance for which the host element should be retrieved.
```
It's not just a component though, it can also be a directive. I named it `target`, because it's consistent with all of the other functions.

IgorMinar replied:
actually, while iterating on this PR, we renamed all "target" arguments to something else and this is the only one remaining how about `componentOrDirective` if you think that `component` is too constraining?

crisbeto commented on tools/public_api_guard/global_utils.d.ts:
```diff
@@ -1,19 +1,24 @@
 export declare function getListeners(element: Element): Listener[];
 
-export declare function getRootComponents(target: {}): any[];
+export declare function getOwningComponent<T>(elementOrDir: Element | {}): T | null;
```
I saw your comment after the meeting so it's too late to bring it up. I have some suggestions, but they aren't much better. We can call it something like `getContainingComponent` or `getClosestComponent`.

IgorMinar replied:
`getContainingComponent` is not too bad. In fact "containing" it's used more frequently in https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.d.ts than `owning`. @kara what do you think?

crisbeto replied:
I ended up renaming it to `getContainingComponent`.

kara replied:
To be honest, I don't love `getContainingComponent` because it's confusing in the case of content children. For ex, below it looks like the containing component is `<comp>`, but it's actually the component that "owns" `<child-comp>` (whatever template `<comp>` and `<child-comp>` are inside of.

```
<comp>
   <child-comp></child-comp>
</comp>
```

I'm good with `getOwnerComponent` though

mary-poppins commented:
You can preview be612f5 at https://pr34453-be612f5.ngbuilds.io/.

crisbeto commented:
@IgorMinar I've addressed/replied to all the recent feedback.

IgorMinar commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -259,46 +279,59 @@ export function getRenderedText(component: any): string {
-  return typeof listener.useCapture === 'boolean';
+  /** Whether the listener is using event capturing. */
+  useCapture: boolean;
+  /** Whether this is a listener for a DOM event. */
```
```suggestion
  /** `true` if this is a listener for a DOM event, `false` if this is an `@Output` listener registered via event binding. */
```

crisbeto commented:
The feedback from last night has been addressed.

mary-poppins commented:
You can preview e33b058 at https://pr34453-e33b058.ngbuilds.io/.

kara commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -24,88 +24,88 @@ import {getTNode, unwrapRNode} from './view_utils';
- *
- * expect(getViewComponent(<child-comp>) instanceof MyApp).toBeTruthy();
- * expect(getViewComponent(<my-app>)).toEqual(null);
+ *   <div>
```
Can we add back the "VIEW" marker here? It's a vital distinction because content children do not work this way. As written, `<child-comp>` would get whatever component `my-app` is inside of.

```suggestion
 * #VIEW
 *   <div>
```

IgorMinar replied:
@kara the dom snippet is supposed to represent what people see and debug in the browser's elements view. There will be no marker there. How can we make this example more clear while writing it from the perspective of a developer debugging the app?

kara replied:
I think in this case having the example from a DOM perspective might be more confusing. Instead, maybe we could clarify with more text?  Something like:

```
For example, if `<child-comp>` is used in the template for `<app-comp>` (i.e. a `ViewChild` of `<app-comp>`), calling `getOwningComponent` on `<child-comp>` would return `<app-comp>`.
```

IgorMinar replied:
sgtm

kara commented on packages/core/src/render3/util/change_detection_utils.ts:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Marks a component for check and synchronously performs change detection on it.
+ *
+ * @param component Component for which to detect changes.
```
This doc makes it sound like we're only performing change detection on the component that's passed in, when in fact we're triggering change detection on the whole application. Can we tweak to reflect that? Maybe:

```
Marks a component for check and synchronously performs change detection on the application.

@param component Component to mark dirty 
```

IgorMinar replied:
as previously discussed, let's not use "dirty" terminology, let's stick to "mark for check" which is already part of our api. But otherwise I like the improvement:

```
Marks a component for check (in case of OnPush components) and synchronously performs change detection on the application this component belongs to.

@param component Component to [mark for check](ChangeDetectorRef#markForCheck).
```

kara replied:
Yeah, that's better

IgorMinar reviewed (changes requested):
@crisbeto Kara has a point about the "containing" terminology. But now I'm confused if she's suggesting to go back to "owner" or "owning"? Let's discuss in the fw sync today and get this done.

kara commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -24,88 +24,88 @@ import {getTNode, unwrapRNode} from './view_utils';
- * otherwise).
- *
- * @param element DOM element which is owned by an existing component's view.
+ * Retrieves the component instance whose view contains the DOM element.
```
```suggestion
 * Retrieves the component instance whose template contains the DOM element.
```

IgorMinar replied:
This seems a bit confusing. The component instance doesn't directly have a template, it has a view and is an instance of a component class that has a template.

kara replied:
Discussed offline, we'll leave it as is

IgorMinar commented:
in an epic bikeshedding session we settled on getOwningComponent because getOwnerComponent has two consecutive nouns.

IgorMinar reviewed (approved):
lgtm after `getContainingComponent` is renamed to `getOwningComponent`

AndrewKushnir commented:
[Presubmit](http://test/OCL:292372356:BASE:292372393:1580407967978:caa364ca)

kara commented on packages/core/src/render3/instructions/change_detection.ts:
```diff
@@ -15,44 +15,34 @@ import {detectChangesInternal, markViewDirty, scheduleTick, tickRootContext} fro
  *
  * @param component The component which the change detection should be performed on.
+ * @publicApi
+ * @globalApi ng
```
`detectChanges` shouldn't be public API. Remove?

IgorMinar replied:
```suggestion
```

IgorMinar commented on packages/core/src/render3/instructions/change_detection.ts:
```diff
@@ -15,44 +15,34 @@ import {detectChangesInternal, markViewDirty, scheduleTick, tickRootContext} fro
+ * This function triggers change detection in a synchronous way on a component.
  *
  * @param component The component which the change detection should be performed on.
+ * @publicApi
```
```suggestion
```

kara reviewed (approved):
LGTM (global)

mary-poppins commented:
You can preview 7c8ce9c at https://pr34453-7c8ce9c.ngbuilds.io/.

crisbeto commented:
Good to go from my end.
